python/cuml/feature_extraction/text.py

```diff
@@ -0,0 +1,576 @@
+        indptr = token_counts.cumsum()
+        indptr = cp.pad(indptr, (1, 0), "constant")
+
+        n_rows = len(doc_token_counts) + len(empty_doc_ids)
+
+        return cp.sparse.csr_matrix(
+            arg1=(data, indices, indptr), dtype=self.dtype,
+            shape=(n_rows, n_features)
+        )
+
+    def _validate_params(self):
+        """Check validity of ngram_range parameter"""
+        min_n, max_m = self.ngram_range
+        msg = ""
+        if min_n < 1:
+            msg += "lower boundary must be >= 1. "
+        if min_n > max_m:
+            msg += "lower boundary larger than the upper boundary."
+        if msg != "":
+            msg = f"Invalid value for ngram_range={self.ngram_range} {msg}"
+            raise ValueError(msg)
+
+    def _warn_for_unused_params(self):
+        if self.analyzer != 'word' and self.stop_words is not None:
+            warnings.warn("The parameter 'stop_words' will not be used"
+                          " since 'analyzer' != 'word'")
+
+
+def _document_frequency(X):
+    """Count the number of non-zero values for each feature in X."""
+    doc_freq = (
+        X[["token", "doc_id"]]
+        .groupby(["token"])
+        .count()
+    )
+    return doc_freq["doc_id"].values
+
+
+def _term_frequency(X):
+    """Count the number of occurrences of each term in X."""
+    term_freq = (
+        X[["token", "count"]]
+        .groupby(["token"])
+        .sum()
+    )
+    return term_freq["count"].values
+
+
+class CountVectorizer(_VectorizerMixin):
+    """Convert a collection of text documents to a matrix of token counts
+
+    If you do not provide an a-priori dictionary then the number of features
+    will be equal to the vocabulary size found by analyzing the data.
+
+    Parameters
+    ----------
+    lowercase : boolean, True by default
+        Convert all characters to lowercase before tokenizing.
+    preprocessor : callable or None (default)
+        Override the preprocessing (string transformation) stage while
+        preserving the tokenizing and n-grams generation steps.
+    stop_words : string {'english'}, list, or None (default)
+        If 'english', a built-in stop word list for English is used.
+        If a list, that list is assumed to contain stop words, all of which
+        will be removed from the input documents.
+        If None, no stop words will be used. max_df can be set to a value
+        to automatically detect and filter stop words based on intra corpus
+        document frequency of terms.
+    ngram_range : tuple (min_n, max_n), default=(1, 1)
+        The lower and upper boundary of the range of n-values for different
+        word n-grams or char n-grams to be extracted. All values of n such
+        such that min_n <= n <= max_n will be used. For example an
+        ``ngram_range`` of ``(1, 1)`` means only unigrams, ``(1, 2)`` means
+        unigrams and bigrams, and ``(2, 2)`` means only bigrams.
+    analyzer : string, {'word', 'char', 'char_wb'}
```

Comment:
Is char_wb actually supported?

Reply:
`char_wb` is the same as `char` when `ngram_range=(1,1)` (which is supported), but when `ngrams > 1` neither `char_wb` nor `char` are supported. There is a [feature request](https://github.com/rapidsai/cudf/issues/5144) to unblock this at cudf.

Reply:
For now if an user tries to use the vectorizer with `analyzer={char|char_wd}` and `ngrams_range=(2,2)` they will get a `NotImplementedError("Character-level ngrams is not yet supported by cuML.")`.

Reply:
`char` and `char_wb` are now supported.